opensquirrel/ir.py

```diff
@@ -82,6 +85,103 @@ def accept(self, visitor: IRVisitor) -> Any:
         return visitor.visit_qubit(self)
 
 
+class Axis(Sequence[np.float64], Expression):
+    """The ``Axis`` object parses and stores a vector containing 3 elements.
+
+    The input vector is always normalized before it is stored.
+    """
+
+    def __init__(self, *axis: AxisLike) -> None:
+        """Init of the ``Axis`` object.
+
+        axis: An ``AxisLike`` to create the axis from.
+        """
+        axis_to_parse = axis[0] if len(axis) == 1 else cast(AxisLike, axis)
+        self._axis = self._parse_and_validate_axislike(axis_to_parse)
+
+    @property
+    def axis(self) -> NDArray[np.float64]:
+        """The ``Axis`` data saved as a 1D-Array with 3 elements."""
+        return self._axis
+
+    @axis.setter
+    def axis(self, axis: AxisLike) -> None:
+        """Parse and set a new axis.
+
+        Args:
+            axis: An ``AxisLike`` to create the axis from.
+        """
+        self._axis = self._parse_and_validate_axislike(axis)
+
+    @classmethod
+    def _parse_and_validate_axislike(cls, axis: AxisLike) -> NDArray[np.float64]:
+        """Parse and validate an ``AxisLike``.
+
+        Check if the `axis` can be cast to a 1DArray of length 3, raise an error
+        otherwise. After casting to an array, the axis is normalized.
+
+        Args:
+            axis: ``AxisLike`` to validate and parse.
+
+        Returns:
+            Parsed axis represented as a 1DArray of length 3.
+        """
+        if isinstance(axis, Axis):
+            return axis.axis
+
+        try:
+            axis = np.asfarray(axis)
+        except (ValueError, TypeError) as e:
+            raise TypeError("Axis requires an ArrayLike") from e
+        axis = axis.flatten()
+        if len(axis) != 3:
+            raise ValueError(
+                f"Axis requires an ArrayLike of length 3, but received an ArrayLike of length {len(axis)}."
+            )
+        return cls._normalize_axis(axis)
+
+    @staticmethod
+    def _normalize_axis(axis: NDArray[np.float64]) -> NDArray[np.float64]:
+        """Normalize a NDArray.
+
+        Args:
+            axis: NDArray to normalize.
+
+        Returns:
+            Normalized NDArray.
+        """
+        return axis / np.linalg.norm(axis)
+
+    def __getitem__(self, index: int, /) -> np.float64:  # type:ignore[override]
+        """Get the item at `index`."""
+        return cast(np.float64, self.axis[index])
+
+    def __len__(self) -> int:
+        """Length of the axis, which is always 3."""
+        return 3
```

Comment:
What do you think of defining 3 (magic number) as a `_len` class member? And then `return _len` from here.

Reply:
Done